### PR TITLE
[Usability] Unify BERT horovod and kvstore pre-training script

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
- changed per GPU batch size option to total batch size option
- add comm_backend option which supports horovod and kvstore (to add byteps support soon)  
- changed short_seq_prob default value to 0

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-889/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/1/index.html

**Comment by mli** [comment]
Job PR-889/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/2/index.html

**Comment by mli** [comment]
Job PR-889/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/3/index.html

**Comment by mli** [comment]
Job PR-889/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/4/index.html

**Comment by mli** [comment]
Job PR-889/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/5/index.html

**Comment by mli** [comment]
Job PR-889/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/7/index.html

File: scripts/bert/run_pretraining.py [hunk 1]
```diff
@@ -31,81 +31,170 @@
...
+                         'Effective only if --raw is set.')
+parser.add_argument('--sp_alpha', type=float, default=1.0,
+                    help='Inverse temperature for probability rescaling for sentencepiece '
+                         'sampling. Effective only if --raw is set.')
+parser.add_argument('--num_data_workers', type=int, default=8,
+                    help='Number of workers to pre-process data. '
+                         'Effective only if --raw is set.')
+parser.add_argument('--eval_use_npz', action='store_true',
+                    help='Set to True if --data_eval provides npz files instead of raw text files')
+parser.add_argument('--backend', type=str, default='device',
```
Comment (apeforest) [inline scripts/bert/run_pretraining.py]: The naming --backend seems not very clear to user. Maybe rename it to "comm_backend"?

File: scripts/bert/run_pretraining.py [hunk 2]
```diff
@@ -31,81 +31,170 @@
...
-                                                              self._nsp_loss, self._vocab_size,
-                                                              args.dtype)
-            ls = ls / self._rescale_factor
+            out = self._model(input_id, masked_id, masked_position, masked_weight,
+                              next_sentence_label, segment_id, valid_length)
+            classified, decoded, ls1, ls2 = out
+            ls = ls1 + ls2
+            ls = ls / args.accumulate
         if args.dtype == 'float16':
             self._trainer.backward(ls)
```
Comment (apeforest) [inline scripts/bert/run_pretraining.py]: This logic is a little bit strange here. Is it true that when args.dtype == 'float16', _trainer is always expected? If so, maybe write `if _trainer is not None:` ?

File: scripts/bert/run_pretraining.py [hunk 3]
```diff
@@ -125,119 +214,167 @@ def train(data_train, model, nsp_loss, mlm_loss, vocab_size, ctx, store):
...
+                    offset = lr * step_num / num_train_steps
+                    new_lr = lr - offset
+                trainer.set_learning_rate(new_lr)
+                if args.profile:
+                    profile(step_num, 10, 14, profile_name=args.profile + str(rank))
+
+            # load data
+            if args.use_avg_len:
+                data_list = [[seq.as_in_context(context) for seq in shard]
+                             for context, shard in zip(ctxs, data_batch)]
```
Comment (apeforest) [inline scripts/bert/run_pretraining.py]: Is this right? it seems len(data_batch) == 1 here.

File: scripts/bert/run_pretraining.py [hunk 4]
```diff
@@ -125,119 +214,167 @@ def train(data_train, model, nsp_loss, mlm_loss, vocab_size, ctx, store):
...
+        target_shape = (args.batch_size, args.dummy_data_len)
+        data_train = get_dummy_dataloader(data_train, target_shape)
+
+    parallel_model = DataParallelBERT(model, trainer=fp16_trainer)
+    num_ctxes = len(ctxs)
     parallel = nlp.utils.Parallel(num_ctxes if num_ctxes > 1 else 0, parallel_model)
 
     while step_num < num_train_steps:
-        for _, dataloader in enumerate(data_train):
+        for _, data_batch in enumerate(data_train):
```
Comment (apeforest) [inline scripts/bert/run_pretraining.py]: Is this correct? where is the `for _, data_batch in enumerate(dataloader):` iteration?

**Comment by mli** [comment]
Job PR-889/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/8/index.html

File: scripts/bert/run_pretraining.py (hunk 4 above)
Comment (eric-haibin-lin) [reply scripts/bert/run_pretraining.py]: the intermediate data iterator is now removed

File: scripts/bert/run_pretraining.py (hunk 3 above)
Comment (eric-haibin-lin) [reply scripts/bert/run_pretraining.py]: should still work if len(ctx) == len(data_batch) == 1

**Comment by mli** [comment]
Job PR-889/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/12/index.html

**Comment by mli** [comment]
Job PR-889/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/13/index.html

**Comment by mli** [comment]
Job PR-889/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/15/index.html

**Comment by mli** [comment]
Job PR-889/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/16/index.html

**Comment by eric-haibin-lin** [comment]
@szha @leezu  Why does renaming the title trigger CI?

**Comment by mli** [comment]
Job PR-889/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/17/index.html

**Comment by leezu** [comment]
It shouldn't, as the PR number (889) doesn't change. None of the recent commits at https://github.com/jenkinsci/github-branch-source-plugin/commits/master seems to (intentionally) introduce this behaviour so I am unsure why it happened.

File: scripts/tests/test_scripts.py [hunk 5]
```diff
@@ -233,127 +234,43 @@ def test_pretrain_create():
...
+                 '--num_steps', '20', '--num_buckets', '1',
+                 '--pretrained',
+                 '--comm_backend', backend]
+
+    if backend == 'horovod':
+        try:
+            # Test only if horovod is present
+            import horovod.mxnet as hvd
+        except ImportError:
+            print("The test expects master branch of MXNet and Horovod. Skipped now.")
```
Comment (leezu) [inline scripts/tests/test_scripts.py]: Can we execute this test on our CI? The time-stamps / execution time for both `master-gpu-integration` and `gpu-integration` suggest that it is skipped for both.

Comment (eric-haibin-lin) [reply scripts/tests/test_scripts.py]: The test with 'device' option should be execute (it's now appearing in the list of slowest tests). I tried to add horovod dependency on CI (https://github.com/dmlc/gluon-nlp/pull/775/files) but I failed with multipled attempts. Unfortunately the log is gone and I cannot find the exact error msg..

File: scripts/bert/run_pretraining.py (hunk 3 above)
Comment (apeforest) [reply scripts/bert/run_pretraining.py]: What I found is that len(ctx) != 1 but len(data_batch) == 1 in my local run. I could retest if this is no longer the case.

**Comment by mli** [comment]
Job PR-889/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/18/index.html

File: scripts/tests/test_scripts.py (hunk 5 above)
Comment (leezu) [reply scripts/tests/test_scripts.py]: Can you remove the try / except block given that horovod is now working on CI? (based on your latest commit and cuda upgrade on CI)

**Comment by mli** [comment]
Job PR-889/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/19/index.html

**Comment by mli** [comment]
Job PR-889/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/20/index.html

**Comment by mli** [comment]
Job PR-889/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/21/index.html

**Comment by mli** [comment]
Job PR-889/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/22/index.html

**Comment by mli** [comment]
Job PR-889/23 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/23/index.html

**Comment by mli** [comment]
Job PR-889/24 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-889/24/index.html

**Pull request merged**